0_2_pandas_dataframes_attributes_methods.Rmd

```diff
@@ -698,13 +826,153 @@ Data Frame.  To get a sample that is half the size the original Data Frame
+
+```{python}
+# a sample with replacement
+df.sample(len(df) + 10, 
+          replace=True) 
+```
+
+Quickly scan the `index` for duplicates, to see that this it definitely contains duplicates.
+
+You task now is the following:
+
+- take a random sample *with replacement* of all the rows in the Dataframe (e.g. your sample should have the *same number of rows* as the original Dataframe)
+
+- use an appropriate method to calculate the means of `Human Development Index`, `Fertility Rate` and `Population`, from your random sample
+
+- using appropriate Pandas methods and indexing operations, calculate the *difference* between the means of `Human Development Index`, `Fertility Rate` and `Population` in your sample vs the means for the same variables in the original data
+
+- store your result in a Pandas Series (note: depending what methods you use, your result may automatically be stored in a Pandas Series)
+
+The result should look something like this (though your differences with almost certainly be different to the ones shown below):
+
+```
+Human Development Index      0.074467
+Fertility Rate              -0.182600
+Population                -210.516580
+Name: mean, dtype: float64
+```
+
+Try to do this in as few lines of code as possible.
+
+```{python}
+# your code here
+```
+
+::: {exercise-end}
+:::
+
+::: {solution-start} more-random
+:class: dropdown
+:::
+
+There are several ways you can achieve this, but one way of doing it in a single line of code is to *chain Pandas methods together*.
+
+The left hand side of the subtraction operation in the cell below uses the `.describe()` method and `.loc` indexing to calculate the means of the original data.
+
+The right hand side of the subtraction draws a sample (the same size as the number of rows in `df`) using `.sample()` with `replace=True`. It then uses `.describe()` to calculate the means of the numeric variables *in this random sample*, and then `.loc` to select just those means (as opposed to including the other information that `.describe()` reports, like standard deviations etc).
+
+The result of this subtraction shows us the difference between the original means and the means of our *resample* (e.g. our random sample, drawn with replacement):
+
+```{python}
+# show the difference in the means (original data vs resample)
+df.describe().loc['mean'] - df.sample(len(df), replace=True).describe().loc['mean']
+```
+
+You can also perform the same process, less efficiently, spread over multiple lines of code. The way shown below is not the only possible way - perhaps you used the `.mean()` method individually on each column, for instance, and then manually constructed a `Series` to store the results (even less efficient!) :
```

Comment:
Or `df.mean()` after dropping the string column?

Reply:
```suggestion
You can also perform the same process, less efficiently, spread over multiple lines of code. The way shown below is not the only possible way - perhaps you used the `.mean()` method individually on each column, for instance, and then manually constructed a `Series` to store the results (even less efficient!) . Alternatively, you might have used `.drop()` to remove non-numeric columns, and then used `.mean()` in parallel on the numeric columns, another valid approach:
```